compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -2026,12 +2026,19 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                 &predicate.subst(tcx, substs),
                 &mut obligations,
             );
-            obligations.push(Obligation {
-                cause: cause.clone(),
-                recursion_depth,
-                param_env,
-                predicate,
-            });
+            if predicate.references_error() {
```

Comment:
Couldn't this check be done here:
https://github.com/rust-lang/rust/blob/a85e94927622665a9e9022de0d33a890a2e32d43/compiler/rustc_trait_selection/src/traits/codegen/mod.rs#L123-L125

It's not obvious that there's any need to prevent errors from propagating in `impl_or_trait_obligations`, whereas turning an ICE into a `delay_span_bug` when `.references_error()`, in the snippet I linked, seems more natural.

But I don't care too strongly about this I don't think.